### PR TITLE
Leaves -- Janice

### PR DESCRIPTION
**Pull request by jaitch**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | It is storing and interacting with the database.
Describe in your own words what the Controller is doing in Rails | It is acting as the go-between among the View, Model and Router.
Describe in your own words what the View is doing in Rails | It is responsible for display functions.
Describe an edge-case controller test you wrote | I tested whether the user will be redirected if they attempt to update a task that doesn't exist.
What is the purpose of using strong params? (i.e. the params method in the controller) | To protect access to the attributes from evil-hearted people, while also obviating the need to tediously type out parameters repeatedly (and inviting in bugs). 
How are Rails migrations related to Rails models? | Rails migrations make changes to the schema of Rails models.
Describe one area of Rails that are still unclear on | I do not feel confident writing Rails tests.

### CODE REVIEW

File: test/controllers/tasks_controller_test.rb [hunk 1]
```diff
@@ -1,126 +1,120 @@
...
+      put toggle_complete_path(task.id)
+      expect(Task.find_by(id: task.id).completion_date).must_equal Date.today
+      must_redirect_to tasks_path
+    end
+
+    it 'goes from the current date to nil' do
+      task = Task.create(name: "take out compost")
+      put toggle_complete_path(task.id)
+      put toggle_complete_path(task.id)
+      expect(Task.find_by(id: task.id).completion_date).must_equal nil
```
Comment (dHelmgren) [inline test/controllers/tasks_controller_test.rb:116]: `DEPRECATED: Use assert_nil if expecting nil from test/controllers/tasks_controller_test.rb:116. This will fail in Minitest 6.
.DEPRECATED: Use assert_nil if expecting nil from test/controllers/tasks_controller_test.rb:58. This will fail in Minitest 6.`

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -1,126 +1,120 @@
...
+      old = Task.count
+      delete task_path(task.id)
+      expect(Task.count).must_equal (old - 1)
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it 'goes from nil to the current date' do
```
Comment (dHelmgren) [inline test/controllers/tasks_controller_test.rb:105]: What if the ID doesn't exist?

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -1,126 +1,120 @@
...
-  # Complete these tests for Wave 4
+  # Tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-
+    it 'decreases Task count by one after deletion' do
+      task = Task.create(name: "take out compost")
+      old = Task.count
+      delete task_path(task.id)
+      expect(Task.count).must_equal (old - 1)
```
Comment (dHelmgren) [inline test/controllers/tasks_controller_test.rb:99]: What if the ID doesn't exist?

File: app/views/tasks/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<h1>To-do:</h1>
+
+<ul>
+  <% @tasks.each do |task| %>
+    <li><strong>
+    <%= link_to task.name, task_path(task) %></strong></li>
+    <br>
+    <%= button_to 'Change Completion Status', toggle_complete_path(task), method: :put, data: {confirm: "Are you sure?"} %>
```
Comment (dHelmgren) [inline app/views/tasks/index.html.erb:8]: why are you having them do a confirmation here? Its not wrong, just a bit weird!

**Comment by dHelmgren** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | yes
Answered comprehension questions | yes
Successfully handles: Index, Show | yes
Index & Show tests pass | yes
Successfully handles: New, Create | yes
New & Create tests pass | yes
Successfully handles: Edit, Update | yes
Tests for Edit & Update test valid & invalid task ids | yes
Successfully handles: Destroy, Task Complete | yes
Tests for Destroy & Task Complete include tests for valid and invalid task ids | see comments
Routes follow RESTful conventions | yes
Uses named routes (like `_path`) |
**Overall** | Good job on Task List! It functions as expected. You hit all of the learning goals: using partials, using strong params, and following Rails conventions.
